### PR TITLE
Make the plugin multi-node aware

### PR DESCRIPTION
**Pull request by matt-richardson**
# Background

Now that TeamCity `2023.05` wants to use the secondary node more (round robin requests), it means that the UI is not showing up on the secondary node.

# Results

I've attempted to make the code multi node aware, but there's not a lot of docs on it.

Also, the way the plugin works with OTEL, we cant really wrap the network calls in the `IOGuard` which they want.

So, what I've done is to only publish OTEL data from the primary node, not the secondary node.

I've tested this out on our preprod instance with builds running solely on primary and solely on secondary, and it seems to work.

Note: this change will make the plugin only work on  TeamCity `2023.05+`. We'll have to do a version bump and call it out in the release notes.

[sc-50898]

# How to review this PR

# Pre-requisites
- [ ] I have considered informing or consulting the right people
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

**Review by cailyoung — APPROVED** [review]
LGTM, however I'm as in the dark as y'all on exactly how this is supposed to work; I couldn't find much documentation out there.

This can be tested out in the preprod TeamCity instance.

**Review by gupta-kartik — APPROVED** [review]
Code LGTM

**Comment by matt-richardson** [comment]
Hmmm.

```

[2023-06-06 02:21:32,516]  ERROR -   jetbrains.buildServer.SERVER - Got unexpected exception, thread "okhttp-dispatch-1" is terminated. Error: java.lang.SecurityException: Connection to "api.honeycomb.io" is prohibited by TeamCity node restrictions

java.lang.SecurityException: Connection to "api.honeycomb.io" is prohibited by TeamCity node restrictions
	at jetbrains.buildServer.serverSide.impl.SecondaryNodeSecurityManager$MySecurityManager.securityException(SecondaryNodeSecurityManager.java:25) ~[server.jar:?]
	at jetbrains.buildServer.serverSide.impl.SecondaryNodeSecurityManager$MySecurityManager.checkConnect(SecondaryNodeSecurityManager.java:53) ~[server.jar:?]
	at java.net.InetAddress.getAllByName0(InetAddress.java:1481) ~[?:?]
	at java.net.InetAddress.getAllByName(InetAddress.java:1385) ~[?:?]
	at java.net.InetAddress.getAllByName(InetAddress.java:1306) ~[?:?]
	at okhttp3.Dns$Companion$DnsSystem.lookup(Dns.kt:49) ~[?:?]
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.kt:169) ~[?:?]
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.kt:131) ~[?:?]
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.kt:73) ~[?:?]
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:205) ~[?:?]
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106) ~[?:?]
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74) ~[?:?]
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255) ~[?:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[?:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[?:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[?:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76) ~[?:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[?:?]
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201) ~[?:?]
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

File: server/src/main/java/com/octopus/teamcity/opentelemetry/server/TeamCityBuildListener.java [hunk 1]
```diff
@@ -53,6 +56,13 @@ public TeamCityBuildListener(
     @Override
     public void buildStarted(@NotNull SRunningBuild build) {
         LOG.debug(String.format("Build started method triggered for %s, id %d", getBuildName(build), build.getBuildId()));
+
+        if (!nodesService.getCurrentNode().isMainNode())
+        {
+            LOG.debug(String.format("This is not the main node - skipping OTEL tracing for %s (id %d) and leaving that responsibility for the main node to deal with", getBuildName(build), build.getBuildId()));
+            return;
```
Comment (cailyoung) [inline server/src/main/java/com/octopus/teamcity/opentelemetry/server/TeamCityBuildListener.java:63]: Is there no way to just not load the build listener at all while retaining the UI extension?

**Review by cailyoung — APPROVED** [review]
LGTM but one question

File: server/src/main/java/com/octopus/teamcity/opentelemetry/server/TeamCityBuildListener.java (hunk 1 above)
Comment (matt-richardson) [reply server/src/main/java/com/octopus/teamcity/opentelemetry/server/TeamCityBuildListener.java:63]: Nice idea - good thinking. 
I just went and investigated, and I dont think there is an easy way.

The DI container (spring) has this defined in xml, so while there's probably a fancy pants way of changing it all up so it only registers if it needs to, that feels like super advanced spring config to me, and that's a whole world of pain that I know very little about.

**Pull request merged**